Upgradeable Statistics stdlib

nalimilan opened the pull request:
This finishes the removal of the Statistics stdlib started at https://github.com/JuliaLang/julia/pull/45594, to make it a separate package. We will then be able to merge Statistics and StatsBase to avoid splitting basic stats functions across multiple modules.

Since there are concerns that users will complain about the need to install a package just to compute the mean and standard deviation, the PR also copies the code from `mean`, `std` and `var` and exports these functions from Base. This is yet one more step in a saga started by https://github.com/JuliaLang/julia/pull/27152. I'm not too happy about this since the choice of functions that live in Base is kind of arbitrary. In particular, its not great that weighted methods defined in StatsBase (to be merged with Statistics) are documented in a different place from the unweighted ones, and we had plans to make weights a keyword argument, which makes dispatching on weights types defined in StatsBase tricky if functions are defined in Base. But well... at least the situation after the PR is less confusing than having both Statistics and StatsBase.

Code is just copied from Statistics, the only changes are:
- add compatibility note to docstrings
- remove references to `Statistics`
- change tests to use the `mean` keyword argument instead of `stdm`/`varm`
- do not test sparse matrices as they are not available in Base

nalimilan committed:
Copy `mean`, `std` and `var` from Statistics

This copies the code for these three methods from the Statistics stdlib module.
Only changes are:
- add compatibility note to docstrings
- remove references to `Statistics`
- change tests to use the `mean` keyword argument instead of `stdm`/`varm`
- do not test sparse matrices as they are not available in Base

nalimilan committed:
Remove Statistics stdlib

nalimilan commented on doc/src/base/statistics.md:
```diff
@@ -0,0 +1,8 @@
+# Statistics
```
Given that this page is very short, we could also put these under the Mathematics section, which already has multiple sections. But it's already quite long.

ViralBShah replied:
We can leave it like this, and it is easy to change later on.

mcabbott commented on base/statistics.jl:
````diff
@@ -0,0 +1,438 @@
+ 2.2285192400943226
+```
+"""
+mean(f, A::AbstractArray; dims=:) = _mean(f, A, dims)
````
If you want to be able to dispatch on a keyword, how strange would it be to build in a hook in for licensed piracy... something like:
```suggestion
mean(f, A::AbstractArray; dims=:, weights=nothing) = _mean(f, A, dims, weights)
```

nalimilan replied:
Yeah that's the kind of pattern I had in mind. But let's leave that question for later as moving Statistics out would already be a big achievement.

DilumAluthge commented:
In my opinion, adding 438 lines of statistics code to Base kind of defeats the point of excising the Statistics stdlib. Because we'll still have 438 lines of code that we can't make breaking changes to, 

For the most common use cases, e.g. `mean(f, arr)`, is it really unreasonable to expect users to just write `sum(f.(arr))/length(arr)` themselves? And for anything more complicated, shouldn't they just install the Statistics package anyway?

There's even less friction for adding packages now, since doing `import Foo` or `using Foo` in the REPL would automatically prompt the user, and they can just press Enter to install the package.

DilumAluthge commented:
Also, what makes `mean`, `std`, and `var` special enough that they need to be included in Base, but other functions don't need to be?

nalimilan commented:
I mostly agree. We could also just remove Statistics for now and see how strongly people complain. As you say, with Pkg improvements, installing an external package now only requires typing `y` after `using Statistics` so it's not a big effort to ask even from newcomers.

That said the previous attempt ended up being reverted (see https://github.com/JuliaLang/julia/issues/27374) so we kind of know that people do complain. The situation may be a bit different now given that even sparse matrix support has been moved out of the stdlib.

bkamins commented:
@nalimilan - thank you for your effort.

Could you please summarize:
1. the design of the ecosystem of Base/Statistics.jl/StatsBase.jl in Julia 1.9 release.
2. the target design of the Base/Statistics.jl/StatsBase.jl ecosystem (if it is different than the one for 1.9 release)
3. How do we expect users to write code that would work under: Julia 1.6 LTS, Julia 1.8, Julia 1.9 and target (if target is different). In particular how would existing code using Statistics.jl and StatsBase.jl be affected?

Thank you!

mschauer commented:
People starting to use `sum(x)/length(x)` could be an indicator of friction caused by moving `mean` to the periphery

bkamins commented:
> People starting to use `sum(x)/length(x)`

This pattern as a replacement of `mean` is plain incorrect as it will not work for `x` that does not have O(1) cost of computing `length` as in such cases it will be undefined.

ViralBShah commented:
mean, var and std are considered common functions, and not special for statistics. People have a very strong objection to needing to install a package for mean, especially. I personally tend to agree with that view myself.

DilumAluthge commented:
I'll concede that `mean(a::AbstractArray)` and `var(a::AbstractArray)` (returning the unbiased sample variance) would be considered common methods to most people. If this PR only added those two methods, I'll withdraw my objection.

But this PR adds a lot of other methods as well. Are all of the methods in this PR really things that most people would consider common? Biased sample variance? Mean of an iterable that doesn't define `length` (or defines it but it's more expensive than constant time)?

I'm a little concerned about the slippery slope here. If we add those methods, then why not add weights? It's not clear to me why some functionality gets to be included, and other functionality is excluded.

DilumAluthge commented:
On a separate note, I'm not sure why we need to add `std(a)` to Base - users can't do `sqrt(var(a))`?

bkamins commented:
While waiting for a comment by @nalimilan about the long-term plan, I think that if we agree `mean([1, 2, 3])` should be included then also e.g. something like `mean(i for i in 1:3 if i > 1.5)` should be included. The cases when `length` is not defined are pretty common.

DilumAluthge commented:
I think I can be talked into adding just `mean(iterable)` and `var(iterable)`, but they support any iterables, including unknown length.

DilumAluthge commented:
Also, I'd be happier with `variance` instead of `var`, but that can be infinitely bikeshed, and that's not the hill I'm going to die on.

ViralBShah commented:
We can bring the weights in as well. I would not be opposed to that.

Numpy actually goes much further in what it considers common functions and has a whole statistics module: https://numpy.org/doc/stable/reference/routines.statistics.html. Now numpy is itself a python package, so it is a little bit apples and oranges.

ViralBShah commented:
BTW, the one thing that I am not sure of is, that with the package manager improvements in the last 4 years, will people have the same reaction to mean not being in Base as they did in 2018. I suspect there will be lesser opposition, but there will still be quite a lot of discontent.

What if Base had, say, `average`, and we then have everything else (mean, std, var) in Statistics?

DilumAluthge commented:
What is `average` here? Just the regular unweighted mean of an array or iterable? E.g. `average(arr::AbstractArray)` or `average(iterable)`?

nalimilan commented:
> Could you please summarize:
>
>     1. the design of the ecosystem of Base/Statistics.jl/StatsBase.jl in Julia 1.9 release.
> 
>     2. the target design of the Base/Statistics.jl/StatsBase.jl ecosystem (if it is different than the one for 1.9 release)
> 
>     3. How do we expect users to write code that would work under: Julia 1.6 LTS, Julia 1.8, Julia 1.9 and target (if target is different). In particular how would existing code using Statistics.jl and StatsBase.jl be affected?

@bkamins As I see it, the goal is to have Statistics as a standalone package to which we will be able to move most or all of the StatsBase features (details to be discussed, we probably don't want to import deprecated API), and to deprecate StatsBase in the long term. For Julia 1.9 probably the only change will be that Statistics is no longer an stdlib module but a separate package, as importing things from StatsBase will take some work.

Nothing should break for users of StatsBase and Statistics, except for having to install Statistics and add `Statistics = "1"` compat version bound to Project.toml (which will be done automatically in the registry like https://github.com/JuliaRegistries/General/pull/66039/commits/c56dc97da98c1cb5c05671837ac500ecded805b9). But since Statistics cannot be upgraded on older Julia versions, packages will likely continue to use StatsBase for a long time (at least until the next LTS is out). Actually it could make sense to have StatsBase merely reexports functions from Statistics on recent Julia versions, a bit like Compat does for Julia Base. The exact implementation still needs reflection, but I don't think we need to sort it out to make a decision regarding this PR.

ViralBShah commented:
I believe JuliaCon 2023 could be the right timeframe for the next LTS.

nalimilan commented:
@DilumAluthge @ViralBShah The reason why I think weighted stats should not be included is that it would pull in a whole machinery of [weight vectors types](https://juliastats.org/StatsBase.jl/latest/weights/) which really cannot live in Base as they are too specialized. That's not a problem for `mean` as we could just have a keyword argument taking any `AbstractVector` (including weight vectors), but for `std` and `var` the correction factor depends on the kind of weight so the types would have to live in Base too.

A better comparison than Numpy is probably [Python's statistics standard library](https://docs.python.org/3/library/statistics.html):
> The module is not intended to be a competitor to third-party libraries such as [NumPy](https://numpy.org/), [SciPy](https://www.scipy.org/), or proprietary full-featured statistics packages aimed at professional statisticians such as Minitab, SAS and Matlab. It is aimed at the level of graphing and scientific calculators.

It is very similar to our current Statistics stdlib, but it also provides mode, harmonic and geometric mean, and basic linear regression. So if we took Python as a reference we would favor the status quo. :-) But the situation isn't ideal in Python as NumPy provides separate definitions, so you have multiple functions to do the same thing.

Conversely, Rust, Go and Swift don't provide any statistical functions in their standard libraries, not even `mean`.

bkamins commented:
@ViralBShah - is changing a module in which some name is defined (`mean`, `mean!`, `var`, and `std` in the case of this PR) considered non-breaking?

mbaz commented:
I use basic statistical functions in my work in signal processing all the time, but I am not a professional statistician. I am not concerned about having to import a package to use `mean`, `var`, `median` and other similar functions. What I am concerned about, though, is the future direction of the Statistics package and the impllications for people like me who just need basic functionality and whose data is mostly represented as regular, plain vectors and matrices.

Recent discussions over in Discourse (see for example [here](https://discourse.julialang.org/t/request-for-unstdlibfication-of-statistics/82443)) make it clear that the desire of the professional statisticians in the community want to adapt Statistics.jl to meet their needs. Among other things, there is a clear reluctance to continue to support functions that take anything other than tables. Such changes would make it difficult and inconvenient for me (and I suspect, many others) to use that package in my own projects.

My concrete proposal is to excise statistics from stdlib if you must, but give a different name to the new package (maybe BasicStats or ClassicStats [or NonSeriousStats :wink: ]). Then the serious statisticians would be free to drive Statistics.jl forward, while the rest of us would continue to use the more basic functions.

ViralBShah commented:
> @ViralBShah - is changing a module in which some name is defined (`mean`, `mean!`, `var`, and `std` in the case of this PR) considered non-breaking?

Perhaps this is more of a @KristofferC or @StefanKarpinski question. I believe that since these are exported names, the module name is less important. And of course, the Statistics.jl package can still provide compatibility.

mschauer commented:
Speaking for the serious statisticians: we have no interest in developing the package in a direction where it stops being practical

nalimilan commented:
> Recent discussions over in Discourse (see for example [here](https://discourse.julialang.org/t/request-for-unstdlibfication-of-statistics/82443)) make it clear that the desire of the professional statisticians in the community is to adapt Statistics.jl to meet their needs. Among other things, there is a clear reluctance to continue to support functions that take anything other than tables. Such changes would make it difficult and inconvenient for me (and I suspect, many others) to use that package in my own projects.

AFAICT this is only what @juliohm proposed. None of the main developers of Statistics and StatsBase have suggested dropping the existing API in favor of another API requiring to use Tables.jl. My goal with this PR is merely to merge Statistics and StatsBase, without changing most of their features.

nalimilan committed:
Fixes

juliohm commented:
> For the most common use cases, e.g. `mean(f, arr)`, is it really unreasonable to expect users to just write `sum(f.(arr))/length(arr)` themselves? And for anything more complicated, shouldn't they just install the Statistics package anyway?

I cannot emphasize enough how @DilumAluthge's statement translates well my feelings about the current status quo. It would be much much much simpler to erase Statistics from Base and let the stats/tables community drive the evolution of these features forward. We have so many talented people driving DataFrames.jl, Tables.jl, StatsAPI.jl, ..., there is enough movement to advance things further in major ways without waiting for Julia release cycles.

> Also, what makes `mean`, `std`, and `var` special enough that they need to be included in Base, but other functions don't need to be?

Also agree with @DilumAluthge here. I don't think these three convenience functions justify all the contortion that stats developers have to make to maintain multiple dependencies. When people argue that `mean` should be part of Base, they mean they want a a one-liner convenience function that is equivalent to sum(x) / length(x) or something similar for iterables. There is so much more one can do with mean/std/var in a statistical context! Did you know for example that GeoStats.jl's mean, std and quantile are aware of geospatial coordinates? There are so many other features that are missing in Base, which will take forever to implement with Julia release cycles.

> We could also just remove Statistics for now and see how strongly people complain.

That is my number one wish @nalimilan 🙏🏽

> What if Base had, say, `average`, and we then have everything else (mean, std, var) in Statistics?

Love this idea as well @ViralBShah ! Just give users alternative convenience functions like `avg` with sensible and basic implementations and move everything else to outside of Base where weights, stats corrections are considered.

> What I am concerned about, though, is the future direction of the Statistics package and the impllications for people like me who just need basic functionality and whose data is mostly represented as regular, plain vectors and matrices.

I don't see how these future directions can affect operations with regular, plain vectors and matrices @mbaz. First you need to make sure that you are talking about the statistical definitions of mean, std, var with weights, corrections etc. Then, you need to ask yourself if any effort operating over rows/columns of tables would necessarily exclude low-level functions for arrays. My understanding is that even if we move forward with tables as first-class objects, there will be low-level functions still that could be exported for users interested in plain arrays. That is super simple and shouldn't affect our decision process here in my opinion.

strickek commented:
I like the idea of simple average in base. It was a big discussion why using statistics is necessery for mean, but so is julia now and it is ok. To put mean, std and var in base would be a step in the wrong direction.

ararslan commented:
> the one thing that I am not sure of is, that with the package manager improvements in the last 4 years, will people have the same reaction to mean not being in Base as they did in 2018.

The package manager has improved greatly but it doesn't actually solve the simplest case here, which is that a new user will fire up the REPL and try to compute a basic quantity like the mean, only to be faced with:
```
julia> mean(x)
ERROR: UndefVarError: mean not defined
Stacktrace:
 [1] top-level scope
   @ REPL[1]:1
```
This doesn't tell you that you _could_ compute a mean if you load a package first, nor which package would provide it. So the improvements to the package manager don't really affect the initial experience, they've just improved the experience once you figure out what you need to do to compute a mean.

> What if Base had, say, average, and we then have everything else (mean, std, var) in Statistics?

This actually seems worse to me, since it would be unclear whether developers should define methods for `mean` or `average`. (The term "average" is also technically ambiguous, as it can also refer to the median or mode depending on the context, though arithmetic mean is obviously most common.) The situation would be somewhat like what Milan noted with separate functions in Python and NumPy that do the same thing.

> There is so much more one can do with mean/std/var in a statistical context! [...] There are so many other features that are missing in Base, which will take forever to implement with Julia release cycles.

Packages can define methods for `mean` that provide all kinds of additional functionality, that's the  ✨ m a g i c ✨  of dispatch. Just because `mean` lives somewhere doesn't mean that all possible methods it could have need to live in the same place. Providing the basic definitions that act on Base types—what Milan has done here—gives the majority of users as much as they need. Why would people be waiting on release cycles? If Base defines a new type and a `mean` method for it in some new release, the `mean` method can be added to Compat alongside the type, as is done for any other Base function.

> is changing a module in which some name is defined (mean, mean!, var, and std in the case of this PR) considered non-breaking?

It is not breaking, see e.g. https://github.com/JuliaLang/julia/pull/35628 for prior art.

> what makes mean, std, and var special enough that they need to be included in Base, but other functions don't need to be?

They're ubiquitous in any discipline that deals with numbers. I think just about everyone learns them in school at a fairly young age. I remember computing standard deviations in my 7th grade math class, and it wasn't an honors class or anything.

ararslan commented on base/statistics.jl:
```diff
@@ -0,0 +1,438 @@
+function mean(f, itr)
+    y = iterate(itr)
+    if y === nothing
+        return Base.mapreduce_empty_iter(f, +, itr,
+                                         Base.IteratorEltype(itr)) / 0
```
```suggestion
        return mapreduce_empty_iter(f, +, itr,
                                    IteratorEltype(itr)) / 0
```
The `Base.` qualifications are no longer necessary since this is inside of Base. (The same applies to a number of other instances in this file.)

KristofferC commented:
> The package manager has improved greatly but it doesn't actually solve the simplest case here, which is that a new user will fire up the REPL and try to compute a basic quantity like the mean, only to be faced with:

We already have ways to hook into errors:

```julia
 Base.Experimental.register_error_hint(UndefVarError) do io, exc
    if exc.var == :mean
        print(io, """\nmean is provided by the Statistics.jl package, run \
                     `using Pkg; Pkg.add(\"Statistics\"); using Statistics` \
                     to install and load it.""")
    end
end
```

```julia
julia> mean
ERROR: UndefVarError: mean not defined
mean is provided by the Statistics.jl package, run `using Pkg; Pkg.add("Statistics"); using Statistics` to install and load it.
```

mschauer commented:
That's cool! Would make sense to open an issue and think about for which functions this would be a good idea to soften landing for newcomers? For example for `I` which often shows up in formulas without having `using LinearAlgebra` in the context.

ararslan commented:
> We already have ways to hook into errors though

TIL, that's very cool. We could even go a step further and do
```julia
Base.Experimental.register_error_hint(UndefVarError) do io, ex
    for package in packages  # just some predefined list of (ex-)stdlibs
        if isdefined(package, ex.var) && isexported(package, ex.var)
            print(io, ex.var, " is provided by the ", package, " package, run ",
                  "`using Pkg; Pkg.add(\"", package, "\"); using ", package,
                  " to install and load it.")
        end
    end
end
```
which avoids needing to determine some list of identifiers to provide this for, though it would require determining a list of packages. From the user's perspective, it'd be somewhat like the experience for `@deprecate_moved`, which was employed heavily back when stuff was first moving out of Base. What I've described here is kind of an aside though; at least using this as you noted for `mean` would address my first point.

That said, I still think Milan's approach in this PR makes sense and would be a net improvement.

> I'd be happier with `variance` instead of `var`

FWIW this seems reasonable if people are worried about Base "stealing" short names (and we could rename `std` to `stddev` analogously) but it's also worth noting that the functions have been around forever under these names, so keeping them as-is seems most convenient.

halleysfifthinc commented:
I think concerns about having to manually install Statistics.jl as a package are overblown. Multiple extremely popular packages directly or transitively depend on Statistics, so it will be installed for most users automatically. So if `mean`, `std`, and `var` are not added to Base, the user experience will be basically unchanged:

```julia
julia> using Statistics # currently/already necessary
 │ Package Statistics not found, but a package named Statistics is available from a registry.  # ╮
 │ Install package?                                                                            # │─ Very unlikely
 │   (@v1.8) pkg> add Statistics                                                               # │
 └ (y/n/o) [y]:                                                                                # ╯

julia> mean(x)
```

I don't think that adding `mean`, `std`, and `var` is necessary to de-stdlib-ify Statistics. (And it would be easy to add them after the fact if there is a large enough outcry.)

GunnarFarneback commented:
> For the most common use cases, e.g. `mean(f, arr)`, is it really unreasonable to expect users to just write `sum(f.(arr))/length(arr)` themselves?

Yes, it's unreasonable. It's considerably more to write and think about when reading the code. More importantly it forces you to repeat yourself (`arr` appears twice), which is never a good design.

I'm one of those who are not happy with having `mean` behind an import and I would be really unhappy if it requires importing a package which adds to latency or introduces periodic breaking updates. I don't want my basic `mean` function to evolve. I want it to be stagnant and unchanging.

ViralBShah commented:
There is no risk of mean evolving (in a breaking way) in that irrespective of what happens, it is still part of the 1.0 api. 

Statistics is no longer in the system image I believe.

GunnarFarneback commented:
Let me be more clear with what I mean.

One of the arguments for moving things out from base/stdlib is to make it easier to improve the functionality and get evolution instead of stagnation. This is generally a good thing but a downside of evolution is that breaking changes are needed from time to time. Semantic versioning is the tool to handle this in a controlled way but even if you're not affected by the breaking changes in the functionality you use, there is still a cost to it. You need CompatHelper to alert you of upgrades and you need to investigate whether the breaking changes affect you. While the ecosystem makes the transition to a new major version there is also a risk that packages get conflicting compat, which causes a variety of inconveniences, especially noticeable when the breaking change happens in a frequently used package.

For the basic `mean`, the balance of those upsides and downsides are very much on the negative side.

juliohm commented:
1. I think that all this discussion around ups and downs of having a single
function name "mean" exported by default is surreal.

2. Pleasing a few people who are just lazy to add a single line "using
Statistics" in the beginning of their scripts instead of advancing the
technology freely is nonsense to me.

3. If people feel that having stability of "mean" today will be their main
bottleneck in the future, they are living an illusion.

4. The argument in favor of stagnation in a community that is constantly
generating innovation and breaking changes is incompatible with the
scientific culture.

Em sex., 2 de set. de 2022 05:35, Gunnar Farnebäck ***@***.***>
escreveu:

> Let me be more clear with what I mean.
>
> One of the arguments for moving things out from base/stdlib is to make it
> easier to improve the functionality and get evolution instead of
> stagnation. This is generally a good thing but a downside of evolution is
> that breaking changes are needed from time to time. Semantic versioning is
> the tool to handle this in a controlled way but even if you're not affected
> by the breaking changes in the functionality you use, there is still a cost
> to it. You need CompatHelper to alert you of upgrades and you need to
> investigate whether the breaking changes affect you. While the ecosystem
> makes the transition to a new major version there is also a risk that
> packages get conflicting compat, which causes a variety of inconveniences,
> especially noticeable when the breaking change happens in a frequently used
> package.
>
> For the basic mean, the balance of those upsides and downsides are very
> much on the negative side.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/JuliaLang/julia/pull/46501#issuecomment-1235228073>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAZQW3IQKHWZDTSXEOZACZ3V4G34VANCNFSM57YDJQ4A>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

mschauer commented:
> Pleasing a few people who are just lazy to add a single line "using
Statistics" in the beginning of their scripts instead of advancing the
technology freely is nonsense to me.

Let's be careful not to let this discussion become too heated and let's keep in mind that these things are compromises balancing different interests between research scientists, software engineers, people in need of a scientific table calculator etc. Nobody is "lazy" here.

ViralBShah commented:
@nalimilan I personally am quite comfortable moving forward with the excision, with mean, std and var having the basic implementations in Base - and extended in Statistics.jl through keyword arguments for weights and such. We also have enough time in the release cycle to see how this all shakes out.

ViralBShah commented:
Bump. Are we waiting on triage to chime in here?

ViralBShah committed:
Merge branch 'master' into nl/statistics

ViralBShah commented:
@KristofferC In order to actually merge anything here, what else would we like to do? Run some Pkgeval?

nalimilan commented:
Yes I was waiting for triage to have a look. I can push a few fixes once it's done.

JeffBezanson commented:
I'd be willing to accept the compromise of putting the `mean` identifier in Base, but unfortunately this seems to be a significant amount of code. Anybody who is in favor of this is not allowed to complain about the size of system images ever again --- unlike stdlibs, this code can't even be excluded in an automated way.

oscardssmith commented on base/statistics.jl:
````diff
@@ -0,0 +1,438 @@
+```
+"""
+function mean(f, itr)
+    y = iterate(itr)
````
`mapreduce(f, +, itr)`

oscardssmith commented:
Triage agree that removing the Statistics stdlib is a good idea. Triage is reluctantly willing to compromise on adding `mean` to Base (although the code should get significant cleanup first), but `var` and `std` really seem pretty statisticsy (and are a lot of code), and we built a really good package manager for a reason.

ViralBShah commented:
The proposal is just the basic `mean`. Weights machinery will be in the package.

bkamins commented:
> The proposal is just the basic `mean`.

But do you mean only `mean(f, itr)` method or also methods for arrays?

nalimilan commented:
I agree that only putting unweighted `mean` in Base would be a reasonable solution. That code is quite short: 50 lines of code excluding docstrings for `mean(itr)`, `mean(f, itr)` and `mean(A::AbstractArray; dims)`. When you say it's long and needs cleanup, do you refer to the weighted mean?

KristofferC commented:
> @KristofferC In order to actually merge anything here, what else would we like to do? Run some Pkgeval?

Version in https://github.com/JuliaStats/Statistics.jl/blob/master/Project.toml needs to be set to 1.9 and merged into Julia.

A PR like https://github.com/JuliaRegistries/General/pull/66039 needs to be created.

Then run PkgEval here with the abovementioned PR as the registry branch.

KristofferC commented:
> A PR like https://github.com/JuliaRegistries/General/pull/66039 needs to be created (I can do that).

https://github.com/JuliaRegistries/General/pull/68391

ViralBShah commented:
> Is it really that big of a problem having `mean` in Statistics if we give a good error message for it..? Even in Python you need to `import statistics` and doing `using Statistics` will query you for automatically installing it so it is roughly the same amount of work.

`numpy` does have `mean` and much more. So you don't really need `statistics` to do the basics. https://numpy.org/doc/stable/reference/generated/numpy.mean.html

bkamins commented:
Having considered the issue:
* me, any many stats people, anyway will do `using Statistics` as one of the first commands when working with Julia
* however, many users are probably happy with having `mean` just as they have `sum` and almost never use `Statistics`

So my vote would be to add `mean` to Base (but I am also OK with having it in `Statistics` only).

KristofferC commented:
> numpy does have mean and much more.

I'm not sure in which way this argues. Like ok, but then you have to install and import numpy?

nalimilan committed:
Drop `std` and `var`, add tests for `mean!`

nalimilan committed:
Merge remote-tracking branch 'origin/master' into nl/statistics

nalimilan commented:
I've updated the PR to drop `std` and `var`. Since having a file and documentation section only for `mean` and `mean!` would be weird, I moved them to already existing files. I also added tests for `mean!` as I noticed that function wasn't tested at all in Statistics.jl (it used to be tested indirectly via `mean`, but it uses a separate implementation for some time!).

Is this good to go then? I don't have a strong opinion about whether `mean` should be included in Base or not, but it would be too bad to block this PR only because we can't make a decision.

ViralBShah commented:
We need a pkgeval run.

nalimilan commented:
@nanosoldier `runtests(ALL, vs = ":master")`

DilumAluthge commented:
@nalimilan Can you update the PR title?

PallHaraldsson commented:
Can we just remove Statistics stdlib and not add anything to Base, or if anything at all, unexported to not commit to it in 1.x+1? Jeff:

>I'd be willing to accept the compromise of putting the mean identifier in Base, but unfortunately this seems to be a significant amount of code. Anybody who is in favor of this is not allowed to complain about the size of system images ever again

I was kind of sympathetic to "mean, std and" in Base, as a compromize for casual users, but I DO want to reserve the right complain and want smaller sysimage. If/since we're just left with mean, even with small extra code in Base, is it worth it and having to explain you need to import for other, std etc, or just for extra keywords of mean?

nanosoldier commented:
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/46501#issuecomment-1257058613) has completed - possible new issues were detected. A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/c5abbc0_vs_e6d9979/report.html).

nalimilan commented:
Ah silly me, we need to merge @KristofferC's PR that registers Statistics as a package first. The registered version also has to reexport `mean` and `mean!` from Base instead of redefining them to avoid conflicts. See https://github.com/JuliaStats/Statistics.jl/pull/128.

nalimilan committed:
Merge branch 'master' into nl/statistics

nalimilan commented:
@nanosoldier `runtests(ALL, configuration=(registry="kc/statistics_compat", ), vs = ":master")`

nalimilan committed:
Add final newlines

nanosoldier commented:
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/46501#issuecomment-1273098849) has completed - possible new issues were detected. A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/d672ebb_vs_6904559/report.html).

ViralBShah commented:
Is it possible this will make it into 1.9 before the feature freeze - in the next few days?

DilumAluthge commented:
Statistics can't be removed until SuiteSparse and SparseArrays are removed (because Statistics depends on SparseArrays).

ViralBShah commented:
I had forgotten, we had removed SparseArrays and then put it back in the system image. The native code caching in 1.9 could perhaps pave the path for all of these things to move out of the system image.

vtjnash commented:
I think we recently fixed the codegen bug that had forced us to put it back in

DilumAluthge commented:
IIRC from an earlier Slack conversation with @KristofferC, we also don't want to remove Statistics/SuiteSparse/SparseArrays from the system image or the standard library until Julia has support for weak dependencies.

ViralBShah commented:
Right, so one way or another if package load times are under control, we can move on this.

Wimmerer commented:
Are we removing Stats/Delimited/SparseArrays in 1.10? Regardless can we separate out the `mean` -> Base part? I hear a lot from Alan about this one :)

nalimilan committed:
Merge branch 'master' into nl/statistics

nalimilan commented:
What's the way forward here? Should I update the PR to make Statistics an "upgradable stdlib" like DelimitedFiles (https://github.com/JuliaLang/julia/pull/48671/files)?

vchuravy commented:
Yeah I think that would be a great step forward!

nalimilan committed:
Make Statistics an stdlib again

nalimilan committed:
Backport JuliaStats/Statistics.jl##131

nalimilan commented:
OK, I've updated to PR to mimic what was done for DelimitedFiles by https://github.com/JuliaLang/julia/pull/45540 and https://github.com/JuliaLang/julia/pull/48671. I'm not sure I got it right since the latter reverts most of the former, but it works. One difference I noticed is that Statistics already isn't precompiled on current master but DelimitedFiles is. Not sure whether it should.

CI fails because of the conflicts with `mean` and `mean!` being exported by both Base and Statistics. Looks like I should update https://github.com/JuliaStats/Statistics.jl/pull/128 to reexport these from Base if they are defined there. Then we will have to update the General registry like https://github.com/JuliaRegistries/General/pull/68391 did.

KristofferC commented:
> One difference I noticed is that Statistics already isn't precompiled on current master but DelimitedFiles is. Not sure whether it should.

I think you need to add it to https://github.com/JuliaLang/julia/blob/master/pkgimage.mk

nalimilan committed:
Adapt pkgimage.mk

nalimilan committed:
Use latest Statistics.jl to fix building manual

nalimilan committed:
Also update checksums

nalimilan committed:
Fix missing ending newline

vchuravy commented on test/reduce.jl:
```diff
@@ -705,3 +705,139 @@ let a = NamedTuple(Symbol(:x,i) => i for i in 1:33),
+    @test_throws DimensionMismatch mean!(zeros(2, 2), x)
+    @test_throws DimensionMismatch mean!(zeros(2, 3), x)
+    @test_throws DimensionMismatch mean!(zeros(5, 2), x)
+end
```
New line at end

vchuravy commented on julia.spdx.json:
```diff
@@ -446,11 +434,6 @@
-            "spdxElementId": "SPDXRef-JuliaStatistics",
-            "relationshipType": "BUILD_DEPENDENCY_OF",
-            "relatedSpdxElement": "SPDXRef-JuliaMain"
-        },
```
I think technically we still want to include Statistics in the spdx since we ship it?

nalimilan replied:
Maybe. I copied what was done for DelimitedFiles at https://github.com/JuliaLang/julia/pull/45540. But when that PR was merged DelimitedFiles was supposed to be removed so maybe it should be added back.

vchuravy commented on NEWS.md:
```diff
@@ -52,6 +52,9 @@ Standard library changes
 
+#### Statistics
+
+* Statistics has been promoted from being a standard library to a separate package. It now has to be explicitly installed to be used. However, functions `mean` and `mean!` are now exported directly by Base. ([#46501])
```
```suggestion
* Statistics is now an upgradeable standard-library. The functions `mean` and `mean!` are now exported directly by Base. ([#46501])
```

vchuravy commented:
Can someone remind me why we are moving `mean`/`mean!` into Base? Already in 1.9 I need to say `using Statistics` to access them?

We are not materially changing Statistics, it is still a standard-library, but users can install now newer versions.

nalimilan committed:
Merge remote-tracking branch 'origin/master' into nl/statistics

nalimilan committed:
Revert moving `mean` to Base

nalimilan committed:
Update NEWS.md

nalimilan commented:
Right, we can probably revert this now. So the PR becomes really minimal! :-) I've pushed a commit to do that.

mschauer commented on NEWS.md:
```diff
@@ -57,6 +57,9 @@ Standard library changes
 
+#### Statistics
+
+* Statistics is now an upgradeable standard library.([#46501])
```
```suggestion
* Statistics is now an upgradeable standard library ([#46501]).
```

vchuravy commented:
So I think the only real thing left to do is step 4 and 5 in https://github.com/JuliaLang/julia/issues/50697

nalimilan commented:
> So I think the only real thing left to do is step 4 and 5 in #50697

Step 4: https://github.com/JuliaRegistries/General/pull/89713
Step 5: https://github.com/JuliaLang/Pkg.jl/pull/3587 (failing until step 4 is completed AFAICT).

> Ah and I see in #45540 we did remove it from the spdx file.

Right, though as I noted the situation was different at the time since the module was completely removed. Maybe we just forgot to revert that.

bkamins commented:
> We will then be able to merge Statistics and StatsBase to avoid splitting basic stats functions across multiple modules.

Just to make sure. With all these updates we are still going to merge Statistics.jl and StatsBase.jl as the last step some time in the future?

nalimilan commented:
I hope so. One complication will be that we probably won't be able to add dependencies to Statistics, which could block porting some features.

bkamins commented:
> which could block porting some features.

That is why I am asking. I think we should have some plan prepared for this. So that we do not end up with a situation where we cannot achieve what we wanted to do in the first place. In particular - where things that cannot go to Statistics.jl should be moved (as I assume leaving leftovers in StatsBase.jl is not a good idea).

If no one has thought about it yet - please let me know and I can try to check.

nalimilan commented:
At any rate making Statistics upgradable is an improvement. But yeah it would be good to list at https://github.com/JuliaStats/Statistics.jl/issues/87 features that rely on dependencies and see what we can do about them. Thanks for offering your help!

cmcaine commented:
If Statistics becomes an upgradeable bundled library (sounds great to me!) then can we leave `mean` out of Base after all? That's less change for end users and seems like less overall confusion to me.

nalimilan commented:
> If Statistics becomes an upgradeable bundled library (sounds great to me!) then can we leave `mean` out of Base after all? That's less change for end users and seems like less overall confusion to me.

Yes, that's what the PR does now.

Wimmerer commented:
I will briefly be the voice of Alan here and say that he detests the removal of `mean`. For what it's worth 😅

I think I hear about it every other meeting

nalimilan committed:
Bump Pkg

nalimilan committed:
Merge remote-tracking branch 'origin/master' into nl/statistics

nalimilan committed:
Bump Statistics

nalimilan commented:
Step 5 (https://github.com/JuliaLang/Pkg.jl/pull/3587) has just been completed, so it looks like we're good?
